docs: GlobalAlloc: completely replace example with one that works

ijackson opened the pull request:
Since this is an example, this could really do with some review from someone familiar with unsafe stuff!

I made the example no longer `no_run` since it works for me.

Fixes #81847

rust-highfive commented:
r? @sfackler

(rust-highfive has picked a reviewer for you, use r? to override)

bluss commented on library/core/src/alloc/global.rs:
```diff
@@ -20,24 +20,51 @@ use crate::ptr;
+///             new = counter;
+///             Some(counter)
+///         }).unwrap();
+///         if new > ARENA { abort(); }
```
The allocator returns null on allocation failure, it does not abort

ijackson replied:
> The allocator returns null on allocation failure, it does not abort

Ideally, yes.  But with the current structure of the critical section (no conditional other than the one inherent in `fetch_update`, the counter would eventually overflow after many many repeated allocations.

Although this is only a demo, I think your implication, that this ought to be addressed, is right, because it's supposed to be demoing the `GlobalAlloc` trait.  I will improve this.

bluss replied:
I would make it limited (in features), complete and correct. Use a Mutex to simplify the code for the reader. Compute the needed bump for the allocation, if there is no more space, return null, otherwise "commit" to it and update and release the lock. Limited could mean that it only supports one alignment (everything is 16-aligned and larger aligns fail to allocate?) or other realistic restriction, just for this example. just to make the example easier to read and write.

ijackson replied:
I like the suggestion of trying to use a Mutex.  Unfortunately they can't be constructed in const context (AIUI because of Windows).  I think `fetch_update` is a good primitive to use but I take your point that I could just bung everything relevant into the critical section which would simplify matters considerably.

The alignment comes out much nicer when counting down so it may turn out to be shorter to keep the alignment support than special case it.

sfackler commented on library/core/src/alloc/global.rs:
```diff
@@ -20,24 +20,51 @@ use crate::ptr;
+/// unsafe impl Sync for SimpleAllocator { }
+///
+/// unsafe impl GlobalAlloc for SimpleAllocator {
+///     unsafe fn alloc(&self, layout: Layout) -> *mut u8 {
```
It seems a bit off that the main ~50% of this example is just bit-twiddling that isn't really relevant to understanding the allocator interface.

It might be better to have a non-compiling example that is easier to understand than a compiling one that has a lot of unnecessary complexity.

ijackson replied:
> It seems a bit off that the main ~50% of this example is just bit-twiddling that isn't really relevant to understanding the allocator interface.

The bit twiddling is a shame, yes.  I was going to say "it's a shame we don't have a method on ptr that does this" but I had overlooked `align_offset`.  So I will change it to use that.

> It might be better to have a non-compiling example that is easier to understand than a compiling one that has a lot of unnecessary complexity.

I reread the API specs for `GlobalAlloc` and `Layout` and there are many subtleties which a complete working example gives an opportunity to demonstrate (and discuss in comments).  Prompted by @bluss's comment I've been through the docs again and I think you'll see what I mean when I push an update to this MR...

ijackson replied:
`align_offset` works on pointers and is fiddly to use.  I took a different approach.

ijackson commented:
I've updated it quite a lot, after review of the `GlobalAlloc` docs.  Now it counts downwards which gets rid of most of the bit-twiddling.  It has more code and comments to handle exceptions and corner cases, to illustrate the subtleties of the API.

ijackson commented:
How about this.  Moving the error handling into the critical section has considerably simplified it.  Almost all the bit-twiddling is gone.  I replaced the special arithmetic operations with plain `-=` etc.

ijackson commented:
I'm not sure the comment for `align_mask_to_round_down` is pulling its weight, but it does discuss an aspect of the `GlobalAlloc` API.  Deleting it might make the code clearer rather than less clear.

bluss commented on library/core/src/alloc/global.rs:
```diff
@@ -20,24 +20,55 @@ use crate::ptr;
+///                 return None
+///             }
+///             remaining -= size;
+///             remaining &= align_mask_to_round_down;
```
There are alignment + size combinations where this is incorrect - too big alignment and we end up with an allocation that overlaps with a previous one

ijackson replied:
Thanks for the detailed review.  I appreciate your critical eye.

> In addition, nothing accounts for the alignment of `self.arena.get()`.

You are entirely right.  I will add `#[repr(C, align( ))]`.

> There are alignment + size combinations where this is incorrect unfortunately - too big alignment and we end up with an allocation is misaligned

This is a consequence of the former.  The construction ensures that the offset is a multiple of the alignment.  If the alignment requested is greater than the alignment of the buffer, this goes wrong.  That can only happen if the alignment is greater than the size which is unusual, but permitted by the API.

> Even as an example, I think the allocator must check all its preconditions and fail the allocation otherwise.

Yes.  Absolutely.

ijackson replied:
Fixed these issues I think.

Dylan-DPC commented:
@sfackler this is ready for review

Dylan-DPC commented:
@sfackler this is ready for review

sfackler commented:
I'm still concerned that the example both:

* is not a demonstration of an actual "real-world" implementation.
* involves a bunch of atomics and bit-twiddling.

If we want to emphasize that alignment requirements must be respected, then we can talk about it in the docs. I don't think we need to force an arbitrarily artificial example just for its own sake.

ijackson commented:
> I'm still concerned that the example both:
> 
>     * is not a demonstration of an actual "real-world" implementation.

I don't agree that this isn't a "real-world" implementation.  I think this implementation would be suitable for an embedded environment where only a small number of allocations will ever occur.

I seriously anticipate that someone might c&p this into their microcontroller.

>     * is 90% atomics and bit-twiddling.

I recognise this is hyperbole but I wanted to pick at it anyway.

The example is 49 lines (l.26 to 74 of the file).  Of those, lines 30, 47, 49-52, and 59 (6 lines in all) would be gone in an implementation which didn't care about or check alignment (of which only lines 49,51 and 59 are actually bit-twiddling).  Lines 27, 62-64 and arguably half of each of 33, 39 and 72 (call that 6 lines in all) would be gone in an implementation which didn't care about threadsafety.

I make that an absolute outside of 12 lines of "atomics and bit-twiddling" - 25%.  I don't think it would be sensible to produce an example which pretended to work but didn't honour alignment, or which wasn't threadsafe, if it saves only a quarter of the size.

The example *is* full of what one might reasonably think of as "weird shit" - the combination of raw pointers, `UnsafeCell` and atomics.  But that comes with the territory.  This "weird shit" density, and the number of available beartraps in that territory, is a big reason why I think a properly worked out example, which gets everything right, is valuable.

> If we want to emphasize that alignment requirements must be respected, then we can talk about it in the docs. I don't think we need to force an arbitrarily artificial example just for its own sake.

I think the best options are something like this example; no example; or a very minimal dummy example - eg as [suggested by @vitalyd](https://github.com/rust-lang/rust/issues/81847#issuecomment-774865544).  The latter is still a fair chunk of text, despite not illuminating much more than how to spell and place the `#[global_allocator]` attribute.  If that's all that's wanted perhaps it could be trimmed down by replacing everything with just calls to `abort` and making it a compile-only test.

If the libs team doesn't feel a proper example is worth the space here then please do feel free to close this MR and leave #81847 open for someone to fix another way.  Obviously that would be disappointing to me but it's not like there aren't other things I could be trying to fix :-).

Alternatively, perhaps this particular example, which demonstrates not just `GlobalAlloc` but also `UnsafeCell`, raw pointers, and atomics, and might well be deployable as-is in some environments even, would be better placed somewhere else.  Then the `GlobalAlloc` docs could perhaps have a reference to it.

Personally ISTM that Rust nowadays likes to put stuff like that in the docs examples, rather than off in some external place which has to be separately hunted down (and isn't tested in CI, etc.).  And it's not like the docs for `GlobalAlloc` are going to be read by Rust newbies.  There's already a large pile of text there and I think a fully worked example is simply commensurate.

Thanks.

JohnTitor commented:
r? @Amanieu for further review

Amanieu commented on library/core/src/alloc/global.rs:
```diff
@@ -20,24 +20,57 @@ use crate::ptr;
-/// unsafe impl GlobalAlloc for MyAllocator {
-///     unsafe fn alloc(&self, _layout: Layout) -> *mut u8 { null_mut() }
-///     unsafe fn dealloc(&self, _ptr: *mut u8, _layout: Layout) {}
+/// const ARENA: usize = 128 * 1024;
```
I feel that `ARENA_SIZE` is a better description of this constant.

Amanieu commented on library/core/src/alloc/global.rs:
```diff
@@ -20,24 +20,57 @@ use crate::ptr;
+///         let align = layout.align();
+///
+///         // Both align==0 and align not power of 2 will mangle remaining (or panic), causing UB.
+///         // This is allowed by the `GlobalAllocator` and `Layout` contracts.
```
Actually the `Layout` contract prohibits creating a layout with an invalid alignment: https://doc.rust-lang.org/nightly/std/alloc/struct.Layout.html#method.from_size_align

ijackson replied:
Yes.  That is what I was trying to say.  I have reworded this since evidently it wasn't clear.  (In my updated branch which I will push shortly.)

Amanieu commented on library/core/src/alloc/global.rs:
```diff
@@ -20,24 +20,57 @@ use crate::ptr;
+///
+///         let mut allocated = 0;
+///         if self.remaining.fetch_update(SeqCst, SeqCst, |mut remaining| {
+///             if size > remaining || align > ARENA { // align may be > size !
```
It might be better to move the `align > ARENA_SIZE` check outside the CAS loop. The compiler can optimize it but it still looks awkward.

ijackson commented:
@Amanieu Thanks for the review comments.  I have addressed them in the top three commits of this branch; I left them as `fixup!` for ease of review, since I have also rebased onto current master.

ijackson commented:
Also I have re-run the example through rustfmt.

Amanieu commented on library/core/src/alloc/global.rs:
```diff
@@ -20,24 +20,69 @@ use crate::ptr;
+///         let align = layout.align();
+///
+///         // `Layout` contract forbids making a `Layout` with align=0, or align not power of 2.
+///         // So we are allowed to write this, even though it can have UB in those cases:
```
```suggestion
///         // So we can safely use a mask to ensure alignment without worrying about UB.
```

Amanieu reviewed (commented):
A minor wording change, but otherwise LGTM. Can you squash the commits before merging?

ijackson commented:
Thanks, folded in your suggestion and squashed.

Amanieu commented:
@bors r+

Amanieu commented:
RIP bors?

@bors r+

Amanieu commented:
@bors r+

bors commented:
:pushpin: Commit 07e11e849586c30540dfd784f56438dc7af11900 has been approved by `Amanieu`

<!-- @bors r=Amanieu 07e11e849586c30540dfd784f56438dc7af11900 -->
<!-- homu: {"type":"Approved","sha":"07e11e849586c30540dfd784f56438dc7af11900","approver":"Amanieu"} -->

bors commented:
:hourglass: Testing commit 07e11e849586c30540dfd784f56438dc7af11900 with merge cf975bbd6418f1571394d60fb875ef5880c24b6e...
<!-- homu: {"type":"BuildStarted","head_sha":"07e11e849586c30540dfd784f56438dc7af11900","merge_sha":"cf975bbd6418f1571394d60fb875ef5880c24b6e"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/3120351822)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/3120351822","builder_name":"checks-actions"} -->

rust-log-analyzer commented:
The job **`i686-msvc-1`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/runs/3120135853) [(plain)](https://github.com/rust-lang-ci/rust/commit/cf975bbd6418f1571394d60fb875ef5880c24b6e/checks/3120135853/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

error: test failed, to rerun pass '--doc'


command did not execute successfully: "\\\\?\\D:\\a\\rust\\rust\\build\\i686-pc-windows-msvc\\stage0\\bin\\cargo.exe" "test" "--target" "i686-pc-windows-msvc" "-Zbinary-dep-depinfo" "-j" "8" "--release" "--locked" "--color" "always" "--features" "panic-unwind backtrace compiler-builtins-c" "--manifest-path" "D:\\a\\rust\\rust\\library/test/Cargo.toml" "-p" "core" "--"


Build completed unsuccessfully in 0:40:40
Build completed unsuccessfully in 0:40:40
make: *** [Makefile:72: ci-subset-1] Error 1
```

</details>

Amanieu commented:
```
2021-07-21T03:30:44.9825882Z ---- src\alloc\global.rs - alloc::global::GlobalAlloc (line 23) stdout ----
2021-07-21T03:30:44.9826661Z unsupported section alignment
2021-07-21T03:30:44.9827523Z UNREACHABLE executed at D:\a\rust\rust\src\llvm-project\llvm\lib\MC\WinCOFFObjectWriter.cpp:297!
2021-07-21T03:30:44.9828342Z Couldn't compile the test.
2021-07-21T03:30:44.9828626Z 
2021-07-21T03:30:44.9828931Z failures:
2021-07-21T03:30:44.9829454Z     src\alloc\global.rs - alloc::global::GlobalAlloc (line 23)
```

`static`s don't support alignments greater than 4k: #70022 #70144

ijackson commented:
> `static`s don't support alignments greater than 4k: #70022 #70144

Oh dear.  I will fix this.  Shouldn't be too hard.

Amanieu commented:
@bors r+

bors commented:
:pushpin: Commit 03d7001564d140a3607a77fe02a5b1aaca0031d6 has been approved by `Amanieu`

<!-- @bors r=Amanieu 03d7001564d140a3607a77fe02a5b1aaca0031d6 -->
<!-- homu: {"type":"Approved","sha":"03d7001564d140a3607a77fe02a5b1aaca0031d6","approver":"Amanieu"} -->
